File: form/wicket/src/main/java/org/opensingular/form/wicket/mapper/richtext/PortletRichTextPanel.js
/*
 * Copyright (C) 2016 Singular Studios (a.k.a Atom Tecnologia) - www.opensingular.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

(function (htmlContainer, hiddenInput, callbackUrl, isEnabled, buttonsList, submitButtonId, classDisableDoubleClick) {

    $(document).ready(function () {
        appendFunctions(window.opener);
    });


    function appendFunctions(opener) {
        $(function () {

            var plugin;
            if (isEnabled === "true") {
                plugin = 'finishAndClose,cancel';
            } else {
                CKEDITOR.config.readOnly = true;
                plugin = 'closed';
            }
            var ids = "";
            var buttonsExtra = buttonsList.split(",");
            buttonsExtra.forEach(function (b) {
                var texts = b.split("#$");
                var id;
                if(texts[3] === "true"){
                    id = 'extra' + texts[0];
                } else {
                    id = texts[0];
                }
                ids += id + ",";
            });
            ids = ids.slice(0, -1);

            var editor = CKEDITOR.replace("ck-text-area", {
                extraPlugins: plugin,
                allowedContent: true,
                skin: 'office2013',
                language: 'pt-br',
                width: '215mm',
                savePlugin: {
                    onSave: function (data) {

                        $('#ck-text-area').val(data);
                        $('#' + submitButtonId).click();
                        var jQuerRefOfHtmlContainer = opener.$('#' + htmlContainer);
                        jQuerRefOfHtmlContainer.html(data);

                        var jQueryRefOfHiddenInput = opener.$('#' + hiddenInput);
                        jQueryRefOfHiddenInput.val(data);
                        jQueryRefOfHiddenInput.trigger("singular:process");
                    }
                },
                toolbar: [
                    {name: 'document', items: ['Closed', 'FinishAndClose', 'Cancel', 'Preview', 'Print']},
                    {
                        name: 'clipboard',
                        items: ['Cut', 'Copy', 'Paste', 'PasteText', 'PasteFromWord', '-', 'Undo', 'Redo']
                    },
                    {name: 'editing', items: ['Find', 'Replace', '-', 'Scayt']},
                    {
                        name: 'basicstyles',
                        items: ['Bold', 'Italic', 'Underline', 'Strike', 'Subscript', 'Superscript', '-', 'RemoveFormat']
                    },
                    {
                        name: 'paragraph',
                        items: ['NumberedList', 'BulletedList', '-', 'Outdent', 'Indent', '-', 'Blockquote', 'CreateDiv', '-', 'JustifyLeft', 'JustifyCenter', 'JustifyRight', 'JustifyBlock']
                    },
                    {name: 'links', items: ['Link', 'Unlink']},
                    {name: 'insert', items: ['Table', 'HorizontalRule', 'SpecialChar', 'PageBreak']},
                    '/',
                    {name: 'styles', items: ['Styles', 'Format', 'FontSize']},
                    {name: 'colors', items: ['TextColor', 'BGColor']},
                    {name: 'tools', items: ['ShowBlocks']},
                    {name: 'others', items: ids.split(",")}
                ],
                on: {
                    'instanceReady': function () {
                        $('.cke_contents').height($('#bodyPage').height() - $('.cke_contents').offset().top - $('.cke_bottom').height() - 20);
                    }
                }
            });

            CKEDITOR.config.disableNativeSpellChecker = false;


            editor.on('doubleclick', function (evt) {
                var element = evt.data.element;
                var classesDoubleClick =  classDisableDoubleClick.split(", ");

                if (element.hasClass(classesDoubleClick)){
                    evt.stop();
                }
            }, null, null, 1);

            buttonsExtra.forEach(function (b) {
                var texts = b.split("#$");

                var id;
                if(texts[3] === "true"){
                    id = 'extra' + texts[0];
                } else {
                    id = texts[0];
                }

                editor.ui.addButton(id,
                    {
                        label: texts[1],
                        command: texts[0],
                        icon: texts[2]
                    });
                editor.addCommand(texts[0], {
                    exec: function () {
                        var selected = editor.getSelection().getSelectedText();
                        var innerText = editor.document.getBody().getText();

                        Wicket.Ajax.post({u: callbackUrl, ep: {'innerText': innerText, 'index': texts[0], 'selected': selected}});

                    }
                });
            });

            return editor;
        });

    }
})('${htmlContainer}', '${hiddenInput}', '${callbackUrl}', '${isEnabled}', '${buttonsList}', '${submitButtonId}', '${classDisableDoubleClick}');

File: form/wicket/src/main/java/org/opensingular/form/wicket/mapper/richtext/PortletRichTextPanel.java
/*
 * Copyright (C) 2016 Singular Studios (a.k.a Atom Tecnologia) - www.opensingular.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.opensingular.form.wicket.mapper.richtext;

import org.apache.wicket.AttributeModifier;
import org.apache.wicket.RestartResponseException;
import org.apache.wicket.markup.ComponentTag;
import org.apache.wicket.markup.head.IHeaderResponse;
import org.apache.wicket.markup.head.OnDomReadyHeaderItem;
import org.apache.wicket.markup.html.WebMarkupContainer;
import org.apache.wicket.markup.html.basic.Label;
import org.apache.wicket.markup.html.form.HiddenField;
import org.apache.wicket.markup.html.link.Link;
import org.apache.wicket.markup.html.panel.Panel;
import org.apache.wicket.model.IModel;
import org.apache.wicket.model.Model;
import org.opensingular.form.wicket.WicketBuildContext;
import org.opensingular.form.wicket.enums.ViewMode;
import org.opensingular.form.wicket.model.SInstanceValueModel;
import org.opensingular.lib.commons.util.Loggable;
import org.opensingular.lib.wicket.util.resource.DefaultIcons;

import java.util.Optional;
import java.util.UUID;

import static org.apache.commons.lang3.StringUtils.EMPTY;

public class PortletRichTextPanel extends Panel implements Loggable {

    private HiddenField<String> hiddenInput;
    private Label label;
    private WicketBuildContext ctx;
    private boolean readOnlyMode;
    private WebMarkupContainer buttonEditar;
    private WebMarkupContainer htmlContent;
    private String previewFrameUuid = UUID.randomUUID().toString();

    @Override
    public void renderHead(IHeaderResponse response) {
        super.renderHead(response);
        response.render(OnDomReadyHeaderItem.forScript(updatePreviewScript()));
    }

    public PortletRichTextPanel(String id, WicketBuildContext ctx, boolean readOnlyMode) {
        super(id);
        this.ctx = ctx;
        this.readOnlyMode = readOnlyMode;
    }

    public WebMarkupContainer configureLabelButton() {
        return new WebMarkupContainer("containerLabel") {

            IModel<String> buttonMsg = new Model<>();
            Label labelMsg = new Label("buttonMsg", buttonMsg);
            WebMarkupContainer iconeClass = new WebMarkupContainer("iconeClass");

            @Override
            protected void onInitialize() {
                super.onInitialize();
                this.add(iconeClass);
                this.add(labelMsg);
            }

            @Override
            protected void onConfigure() {
                super.onConfigure();
                if (!isReadOnlyMode()) {
                    buttonMsg.setObject("Editar");
                    iconeClass.add(new AttributeModifier("class", DefaultIcons.PENCIL));
                } else {
                    buttonMsg.setObject("Visualizar");
                    iconeClass.add(new AttributeModifier("class", DefaultIcons.EYE));
                }

            }
        };
    }

    @Override
    protected void onInitialize() {
        super.onInitialize();
        build(ctx);
    }

    private String updatePreviewScript() {
        String script = "(function(){";
        script += "$('#" + htmlContent.getMarkupId(true) + "').html(\"<iframe ";
        script += " id=\\\"" + previewFrameUuid + "\\\" style=\\\"height: 100%; width: 100%;\\\" frameborder=\\\"0\\\"></iframe>\");";
        script += " var frame = $('#" + previewFrameUuid + "')[0].contentWindow";
        script += " || $('#" + previewFrameUuid + "')[0].contentDocument.document";
        script += " || $('#" + previewFrameUuid + "')[0].contentDocument;";
        script += " frame.document.open();";
        script += " frame.document.write( $(\"#" + hiddenInput.getMarkupId(true) + "\").val());";
        script += " frame.document.close();";
        script += "}());";
        return script;
    }

    private void build(WicketBuildContext ctx) {
        label = new Label("label", Model.of(Optional.ofNullable(ctx.getCurrentInstance().asAtr().getLabel()).orElse(EMPTY)));
        hiddenInput = new HiddenField<>("hiddenInput", new SInstanceValueModel<>(ctx.getModel()));
        hiddenInput.setEnabled(ctx.getViewMode() == ViewMode.EDIT);
        htmlContent = new WebMarkupContainer("htmlContent");

        add(label);
        add(hiddenInput);
        add(htmlContent);
        buttonEditar = createButtonOpenEditor();
        buttonEditar.add(configureLabelButton());
        add(buttonEditar);
    }

    private WebMarkupContainer createButtonOpenEditor() {
        return new Link<String>("button") {

            @Override
            protected void onConfigure() {
                super.onConfigure();
                this.setVisible(PortletRichTextPanel.this.isEnabledInHierarchy());
            }

            @Override
            public void onClick() {
                throw new RestartResponseException(new RichTextNewTabPage(label.getDefaultModelObject().toString(),
                        isReadOnlyMode(), PortletRichTextPanel.this.ctx, hiddenInput, previewFrameUuid));
            }

            @Override
            protected void onComponentTag(ComponentTag tag) {
                super.onComponentTag(tag);
                tag.put("onclick", "window.open('" + getURL() + "', '_blank" + previewFrameUuid + "');");
            }

        };
    }

    public boolean isReadOnlyMode() {
        return readOnlyMode || !this.isEnabledInHierarchy();
    }
}